### PR TITLE
Show loading indicator and placeholder when an amp-list refetches

### LINKED ISSUE #13718: amp-list should show loading indicator and placeholder during refetch
When an amp-list refetches its data due to the src changing via amp-bind or the soon-to-be-implemented refresh action being triggered (#13717), the amp-list just shows the previous data until the new data is returned by the server. This can create a weird user experience in lots of cases, because the user taps some button which should change the data but the page seems to not react at all until the refetch is complete. I think it would be better to instead hide the previous data during the refetch and show the loading indicator and placeholder while it is loading to make it clear that the data is being refreshed. 

cc @choumx @aghassemi

### PR DESCRIPTION
**Pull request by josh313**
- Shows the loading indicator and placeholder when an amp-list refetches due to an amp-bind src change or the refresh action (#13725)
- Resolves #13718 

cc @aghassemi @choumx

### CODE REVIEW

File: src/custom-element.js [hunk 1]
```diff
@@ -1488,21 +1489,21 @@ function createBaseCustomElementClass(win) {
...
-              this.signals_.get(CommonSignals.RENDER_START))) {
-        // Loading has already been canceled. Ignore.
-        return;
-      }
       this.loadingState_ = state;
       if (!state && !this.loadingContainer_) {
         return;
       }
 
+      if (state && !this.implementation_.doesReuseLoadingIndicator() &&
```
Comment (aghassemi) [inline src/custom-element.js]: any reason this block moved down?

File: src/base-element.js [hunk 2]
```diff
@@ -695,6 +695,25 @@ export class BaseElement {
...
+   * be called inside a mutate context.
+   * @param {boolean} state
+   * @public @final
+   */
+  toggleLoading(state) {
+    this.element.toggleLoading(state);
+  }
+
+  /**
+   * Returns is reused again after the render.
```
Comment (aghassemi) [inline src/base-element.js]: Returns whether loading indicator is reused...

File: extensions/amp-list/0.1/amp-list.js [hunk 3]
```diff
@@ -145,6 +133,11 @@ export class AmpList extends AMP.BaseElement {
     }
   }
 
+  /** @override */
+  doesReuseLoadingIndicator() {
```
Comment (aghassemi) [inline extensions/amp-list/0.1/amp-list.js]: I like to suggest renaming to `isLoadingReused`. `does`prefix is not commonly used in AMP and other related loading-related methods just say `loading` instead of `loadingIndicator`

File: extensions/amp-list/0.1/amp-list.js [hunk 4]
```diff
@@ -197,7 +196,23 @@ export class AmpList extends AMP.BaseElement {
       return this.scheduleRender_(items);
     }, error => {
       throw user().createError('Error fetching amp-list', error);
-    });
+    })
+        .then(() => {
```
Comment (aghassemi) [inline extensions/amp-list/0.1/amp-list.js]: nit: no new line needed here

File: extensions/amp-list/0.1/amp-list.js [hunk 5]
```diff
@@ -145,6 +133,11 @@ export class AmpList extends AMP.BaseElement {
     }
   }
 
+  /** @override */
```
Comment (aghassemi) [inline extensions/amp-list/0.1/amp-list.js]: a bit of comment to clarify what's happening here, something along the lines of `amp-list reuses the loading indicator when list is fetched again via bind mutation or refresh action`

File: extensions/amp-list/0.1/amp-list.js [hunk 6]
```diff
@@ -177,6 +170,12 @@ export class AmpList extends AMP.BaseElement {
     if (!this.element.getAttribute('src')) {
       return Promise.resolve();
     }
+    this.togglePlaceholder(true);
```
Comment (aghassemi) [inline extensions/amp-list/0.1/amp-list.js]: This feature is technically is breaking UI change and I can see cases where one may not want this feature at all. I recommend putting this behind an attribute `show-loading-on-refresh` or similar. I will bring it up in our design review tomorrow morning to see what others think.

File: src/custom-element.js (hunk 1 above)
Comment (josh313) [reply src/custom-element.js]: Not that I remember. I'm not sure how this got moved. I moved it back now.

File: extensions/amp-list/0.1/amp-list.js (hunk 6 above)
Comment (josh313) [reply extensions/amp-list/0.1/amp-list.js]: Makes sense. Let me know the outcome of the design review and I'll update the code accordingly.

Comment (aghassemi) [reply extensions/amp-list/0.1/amp-list.js]: Added to https://github.com/ampproject/amphtml/issues/13718#issuecomment-371205208.

tl;dr: `reset-on-refresh` attribute needed as opt-in signal to get this behaviour.

File: extensions/amp-list/0.1/amp-list.js [hunk 7]
```diff
@@ -204,6 +208,21 @@ export class AmpList extends AMP.BaseElement {
       return this.scheduleRender_(items);
     }, error => {
       throw user().createError('Error fetching amp-list', error);
+    }).then(() => {
+      if (this.getFallback()) {
+        // Hide in case fallback was displayed for a previous fetch.
+        this.toggleFallbackInMutate_(false);
+      }
+      this.togglePlaceholder(false);
```
Comment (aghassemi) [inline extensions/amp-list/0.1/amp-list.js]: this is a change from previous code. These should happen automatically anyway. I recommend keeping the old code structure to eliminate any regressions e.g.
```
const fetch =...
fetch.then() {
 if(fallback) etc...
}
return fetch;
```

File: extensions/amp-list/0.1/amp-list.js [hunk 8]
```diff
@@ -204,6 +208,21 @@ export class AmpList extends AMP.BaseElement {
...
+      this.toggleLoading(false);
+    }, error => {
+      this.toggleLoading(false);
+      if (this.getFallback()) {
+        this.toggleFallbackInMutate_(true);
+        this.togglePlaceholder(false);
+      } else {
+        throw error;
```
Comment (aghassemi) [inline extensions/amp-list/0.1/amp-list.js]: ditto, if we stick with the old code structure, this should not be needed.

File: extensions/amp-list/validator-amp-list.protoascii [hunk 9]
```diff
@@ -46,6 +46,7 @@ tags: {  # <amp-list>
   attrs: { name: "items" }
   attrs: { name: "max-items" }
   attrs: { name: "single-item" }
+  attrs: { name: "reset-on-refresh" }
```
Comment (aghassemi) [inline extensions/amp-list/validator-amp-list.protoascii]: please add `value=""` that would ensure `reset-on-refresh` can not be given a value and is treated as a binary flag.
also alphabetical ordering of the attributes names

File: extensions/amp-list/0.1/amp-list.js (hunk 7 above)
Comment (josh313) [reply extensions/amp-list/0.1/amp-list.js]: You mean move this `.then(...)` block back to `layoutCallback()`? That doesn't work because then the loading and placeholder don't go away after a reload (either from the refresh action or an amp-bind change). I just tried it now.

Comment (aghassemi) [reply extensions/amp-list/0.1/amp-list.js]: I see

**Comment by aghassemi** [comment]
@josh313 PR LGTM, I am gonna do some manual edge case testing and then merge it.

**Comment by aghassemi** [comment]
@josh313 Meanwhile, could you please take a look at the failing test?

**Comment by josh313** [comment]
@aghassemi Thanks. I fixed the unit tests now. The integration tests appear to have just stalled. I'm guessing that's unrelated to my change. Is there a way to restart it?

**Comment by aghassemi** [comment]
@josh313 failure was due some whitespace nits in the validator test file. added a fix, they should pass now. I will let you know when I finish testing and merging this, will happen this week for sure.

**Comment by josh313** [comment]
@aghassemi Sounds great. Thanks!

**Comment by aghassemi** [comment]
Rebased and did some local testing, looks good. will merge and do more testing on production sites with Canary that's cut tomorrow.

**Comment by yajiv185** [comment]
@aghassemi  From when we will be able to use this feature?

**Comment by aghassemi** [comment]
@yajiv185 It will be in dev channel tomorrow so you can play with the feature, having your AMP pages validate with this feature is ~ 2-3 weeks away.

**Comment by Gregable** [comment]
These validator changes will be live everywhere within an hour.

**Comment by yajiv185** [comment]
@Gregable I think changes related to this feature is not live, right?(because still I getting error 'The attribute 'reset-on-refresh' may not appear in tag 'amp-list'.')
@aghassemi  I am still not able to play with this feature.(am I missing something that I have do before using this feature?)
If is it possible then please you inform here once it will be live(with validateor).
